runtime/lua/vim/diagnostic.lua

```diff
@@ -0,0 +1,1377 @@
+---            - namespace: (number) Limit diagnostics to the given namespace
+---            - severity: (|vim.diagnostic.severity|) Only return diagnostics with this
+---                        severity. Overrides severity_limit
+---            - severity_limit: (|vim.diagnostic.severity|) Limit severity of diagnostics
+---                            found. For example, `vim.diagnostic.severity.WARN` means
+---                            errors and warnings are valid.
+---            - show_header: (boolean, default true) Show "Diagnostics:" header
+---@param bufnr number|nil The buffer number. Defaults to the current buffer.
+---@param position table|nil The (0,0)-indexed position. Defaults to the current cursor position.
+---@return table {popup_bufnr, win_id}
+function M.show_position_diagnostics(opts, bufnr, position)
+  vim.validate {
+    opts = { opts, 't', true },
+    bufnr = { bufnr, 'n', true },
+    position = { position, 't', true },
+  }
+
+  opts = opts or {}
+  opts.focus_id = "position_diagnostics"
+  bufnr = get_bufnr(bufnr)
+  if not position then
+    local curr_position = vim.api.nvim_win_get_cursor(0)
+    curr_position[1] = curr_position[1] - 1
+    position = curr_position
+  end
+  local match_position_predicate = function(diag)
+    return position[1] == diag.lnum and
+    position[2] >= diag.col and
+    (position[2] <= diag.end_col or position[1] < diag.end_lnum)
+  end
+  local position_diagnostics = vim.tbl_filter(match_position_predicate, M.get(opts.namespace, bufnr))
+  if opts.severity then
+    position_diagnostics = filter_to_severity_limit(opts.severity, position_diagnostics)
+  elseif opts.severity_limit then
+    position_diagnostics = filter_by_severity_limit(opts.severity_limit, position_diagnostics)
+  end
+  table.sort(position_diagnostics, function(a, b) return a.severity < b.severity end)
+  return show_diagnostics(opts, position_diagnostics)
+end
+
+--- Open a floating window with the diagnostics from the given line.
+
+---@param opts table Configuration table. See |vim.diagnostic.show_position_diagnostics()|.
+---@param bufnr number|nil The buffer number. Defaults to the current buffer.
+---@param lnum number|nil The line number. Defaults to line number of cursor.
+---@return table {popup_bufnr, win_id}
+function M.show_line_diagnostics(opts, bufnr, lnum)
+  vim.validate {
+    opts = { opts, 't', true },
+    bufnr = { bufnr, 'n', true },
+    lnum = { lnum, 'n', true },
+  }
+
+  opts = opts or {}
+  opts.focus_id = "line_diagnostics"
+  lnum = lnum or (vim.api.nvim_win_get_cursor(0)[1] - 1)
+  local line_diagnostics = M.get_line_diagnostics(bufnr, lnum, opts)
+  return show_diagnostics(opts, line_diagnostics)
+end
+
+--- Clear diagnotics and diagnostic cache.
+---
+---@param namespace number
+---@param bufnr number|nil Reset diagnostics for the given buffer. If omitted,
+---             diagnostics are reset for all buffers.
+function M.reset(namespace, bufnr)
```

Comment:
`clear()` simply removes decorations while `reset()` also actually removes diagnostics. So `reset()` could be replaced with `set(ns, {})`. I'm not sure how useful it is to expose `clear()` publicly, since in practice it's basically the same as `disable()`.

Reply:
One useful part of `reset()` is that it can be used without a `bufnr` argument to remove all diagnostics managed by a particular namespace. This can't be done with `set()`.